std/encoding.d

```diff
@@ -2250,24 +2138,27 @@ body
             immutable minReservePlace = 6;
         }
 
-        Dst[] buffer = new Dst[s.length];
```

Comment:
Compared to idup, it allocates less, but it switches from a single array copy to a series of small array copies. This is a more predictable cost, though not necessarily a smaller one.

Compared to the previous code, it's a series of small slice copies. Shouldn't be terribly expensive.

Most of std.encoding would be pure, but it's using an interesting (and moderately annoying) system to minimize the amount of duplicate code. That system obscures which cases can be pure. Switching functions to no-arg template functions will make some things just work, but encoding to an array, for instance, will take some redesign. And encoding to an array is the case we need here.